lm_eval/models/openai_completions.py

```diff
@@ -527,3 +530,201 @@ def loglikelihood(self, requests):
 
     def loglikelihood_rolling(self, requests):
         raise NotImplementedError("No support for logits.")
+
+
+@register_model("local-chat-completions")
+class LocalChatCompletionsLM(OpenaiChatCompletionsLM):
+    def __init__(
+        self,
+        model: str = None,
+        base_url: str = None,
+        truncate: bool = False,
+        batch_size=None,
+    ) -> None:
+        """
+        Implements an OpenAI-style chat completion API for locally-hosted models using HuggingFace Tokenizer
+        in place of tiktoken
+        Pass in the model name (if Hugging-Face style naming) and tokenizer. via cli args
+        e.g,
+        lm_eval --model local-chat-completions --tasks gsm8k --model_args model=facebook/opt-125m,base_url=http://{yourip}:8000/v1
+
+        :param model: str
+            Local model (Using HuggingFace model paths, e.g. facebook/opt-125m)
+        :param truncate: bool
+            Truncate input if too long (if False and input is too long, throw error)
+        """
+        super(OpenaiChatCompletionsLM, self).__init__()
+        self.model = model
+        self.base_url = base_url
+        assert self.base_url, "must pass `base_url` to use Local LM!"
+        self.frequency_penalty = 0
```

Comment:
I know `self.frequency_penalty` through to `self.top_p` are also done this way in OpenAIChatCompletions, but we should probably either 1) allow these to be set by a user through arbitrary **kwargs passed to the `oa_chat_completions()` function or 2) allow them to be specified via `--gen_kwargs` which might be even more preferable.

Doesn't necessarily have to happen in this PR, but making a note of it here.

Reply:
I can try to implement this in this PR - should I follow the huggingface class for examples of how gen_kwargs work? I'm not seeing an entrypoint for them, but I do see the option to at least pass kwargs here: https://github.com/EleutherAI/lm-evaluation-harness/blob/9e03d9d024be9bc3e92f8c63b5595c1e12c119da/lm_eval/models/huggingface.py#L101

Reply:
Yup! You can safely assume when you unpack a request's `args` attribute it'll be of the form `context, gen_kwargs` where gen_kwargs is a dict containing (the `generation_kwargs` specified by a task - then updated with a dict of the generation kwargs passed via the CLI).

In this case, I'd do something like 
- remove the manual setting of these generation-related attributes in the class init
- replace the `oa_chat_completion` call with

```
 response = oa_chat_completion(
                        client=self.client,
                        messages=inps,
                        model=self.model,
                        **gen_kwargs,
                    )
 ```
and setting the appropriate defaults/fallbacks for `frequency_penalty`, `logit_bias`, `n`, `presence_penalty`, `temperature`, `top_p` plus ideally erroring out or warning if some gen_kwarg (say, do_sample=True) that is invalid for OpenAI is passed by the user.

Reply:
Thanks for the detailed, notes, this was really helpful. I implemented everything except the error (https://github.com/EleutherAI/lm-evaluation-harness/pull/1174/files#diff-390f332b33418a36b2034271770d49b8851cabfafd91998e6b247b1431d9cc7cR527) because I wasn't sure which edge cases we want to cover from the completions class variable list and thought this might get big enough to pull into its own small method for verification/unit testing. What do you think? https://github.com/openai/openai-python/blob/3ad4e8bc9d89d7a81586bf598289ff62b0a339b9/src/openai/resources/chat/completions.py#L62 Happy to remove to the TODO for now or leave it as a placeholder, let me know.

Reply:
Re: edge cases, I think we may want to do `if "do_sample" in kwargs:\n\tkwargs.pop("do_sample")`, (This is an arg HF requires, but other APIs pretty much don't.) but for the rest, I think we should not validate and should allow the OpenAI api to error out when those reach the call to `openai.chat.completions`.

The current version of this PR now ignores `kwargs` passed along with requests and only uses `self.gen_kwargs` though--I think we should do the opposite if possible, and just have

```
class OpenaiChatCompletionsLM(LM):
    def __init__(
        self,
        model: str = "gpt-3.5-turbo",  # GPT model or Local model using HuggingFace model paths
        base_url: str = None,
        truncate: bool = False,
        revision: Optional[str] = "main",
        trust_remote_code: Optional[bool] = False,
        use_fast_tokenizer: Optional[bool] = True,
        **kwargs,
    ) -> None:
```
as the signature of __init__(), not set `self.gen_kwargs` in the __init__(), and pass the `**kwargs` (the variable defined on L511, not the kwargs from the __init__()) to `oa_chat_completions` instead.

Hope this makes sense!

Reply:
Yup, thanks! That makes complete sense since I was puzzling on the best way to include gen_kwargs but didn't think that we wanted kwargs as well so I overspecified.